### ISSUE TITLE
allow text-references and variables anywhere in filter a expression

### ISSUE THREAD

**Issue by tobibeer** [Jermolene/TiddlyWiki5#1427]
Truly, anywhere. Simply apply the text-references and variables first and then run the filter.

Why would they be restricted to operands only?
(Please, don't say because it would be too easy to misconstruct the filter!)

**example:** `<$list filter="{foo!!bar}<baz>{mumble}[each[frotz]<gronk>]"/>`

With some **debug** macro it should be easy to verify if you actually get what you wanted.

**Comment by Jermolene** [comment]
I think there would be some ambiguity; how would `[{foo!!bar}{blah!!bleep}]` be interpreted?

**Comment by tobibeer** [comment]
> [{foo!!bar}{blah!!bleep}]

Simply substitute the text-references, then evaluate the resulting filter expression.

**Comment by Jermolene** [comment]
> Ambiguity, How? Simply substitute the text-references or variables, then evaluate the resulting filter expression.

Bear in mind that filters have a separate parsing, compilation and execution phases. It sounds like your proposed substitutions would be done as part of the parsing. That's problematic because it would mean that we'd need to reparse the filter if the referenced tiddlers change.

**Comment by Jermolene** [comment]
Also, you can already use arbitrary textual substitution to build a filter expression by using a macro; that is rather what macros are for.

**Comment by tobibeer** [comment]
> That's problematic because it would mean that we'd need to reparse the filter if the referenced tiddlers change.

Doesn't the currently limited support, perhaps using macros, yield precisely that, too? I see no difference. The only difference is when the substitution is taking place.

> Also, you can already use arbitrary textual substitution to build a filter expression by using a macro; that is rather what macros are for.

Instead of a simple `<$list filter="{!!foo}<bar>{!!baz}"/>` we are required to literally **develop**...

```
\define getFrotz()
$(foo)$$(bar)$$(baz)$
\end

<$set name="foo" value={{!!foo}}>
<$set name="baz" value={{!!baz}}>
<$set name="mumble" value="getFrotz">
<$list filter=<<mumble>>/>
</$set>
</$set>
</$set>
```

Atm, I fail to see what makes that necessary.

**Comment by Jermolene** [comment]
> Doesn't the currently limited support, perhaps using macros, yield precisely that, too?

Well, macros are the mechanism by which we expose the ability to parse text with textual substitutions. The text substitutions available to macros don't include transclusions of tiddler fields precisely because that would introduce this potential need for the macro to be refreshed when a tiddler changes.

> Instead of a simple `<$list filter="{!!foo}<bar>{!!baz}"/>` we are required to literally develop...

I still don't think that that is simple: if I understand what you mean, it's not clear whether `{!!foo}` is the operand of an operator or the name.

> Atm, I fail to see what makes that necessary.

I think all you're wanting is a briefer syntax for variable assignment, which could well be a good idea.

**Comment by tobibeer** [comment]
>  if I understand what you mean, it's not clear whether {!!foo} is the operand of an operator or the name

Sure, that may not be clear just from looking at this placeholder filter expression. But does it have to be? Whether we use the currently "complicated way" or this "simple way", way we may end up with invalid expressions either way. Those fields might be populated with well established predefined options from a **$select**, though, even constructing a variable number of filter runs, each optional ...however: to be concatenated.

> I think all you're wanting is a briefer syntax for variable assignment, which could well be a good idea.

That would be equally appreciated.
In this case, it would simplify the additional step to...

```
<$set name="evaluated" value="{{!!foo}}" value="<<bar>>" value="{{baz}}">
<$list filter=<<evaluated>>/> 
</$set>
```

My question is, how is that more conclusive than directly doing that with the filter parameter?

I assume the problem with implementing these "compound attributes" is the exceptional handling you suspect for some **filter** / **$filter** parameters only. So, you may want to avoid that to keep widget attributes type-agnostic.

Perhaps, the opposite is what actually helps so as to indicate some explicit processing, e.g....

```
<$list filter=:"{!!foo}<bar>{!!!baz}"/>
```

In other words, evaluate everything after the colon following the equals sign using the special filter replacing modalities and then use the return value as the attribute. This would not work with wikitext macros, but then we do have $macrocall for more advanced evaluations already.

**Comment by pmario** [comment]
New filter operators and improving existing filters tend to get several parameters instead of using "hardcoded" filter suffixes. Filter parameters can access everything. 

Also new functions help a lot. 

So close this one.
